src/inference/hmc.jl

```diff
@@ -221,245 +290,105 @@ end
 ####
 
 # Sampler(alg::Hamiltonian) =  Sampler(alg, AHMCAdaptor())
-function Sampler(alg::Hamiltonian, s::Selector=Selector())
-    info = Dict{Symbol, Any}()
-
-    info[:eval_num] = 0
-    info[:i] = 0
-
-    Sampler(alg, info, s)
-end
-
-
-function sample(
+function Sampler(
+    alg::SamplerType,
     model::Model,
-    alg::Hamiltonian;
-    save_state=false,                                   # flag for state saving
-    resume_from=nothing,                                # chain to continue
-    reuse_spl_n=0,                                      # flag for spl re-using
-    adaptor=AHMCAdaptor(alg),
-    init_theta::Union{Nothing,Array{<:Any,1}}=nothing,
-    rng::AbstractRNG=GLOBAL_RNG,
-    discard_adapt::Bool=true,
-    verbose::Bool=true,
-    progress::Bool=false,
-    kwargs...
-)
-    # Create sampler
-    spl = reuse_spl_n > 0 ? resume_from.info[:spl] : Sampler(alg)
-    @assert isa(spl.alg, Hamiltonian) "[Turing] alg type mismatch; please use resume() to re-use spl"
-
-    # Resume selector
-    resume_from != nothing && (spl.selector = resume_from.info[:spl].selector)
-
-    # TODO: figure out what does this line do
-    n = reuse_spl_n > 0 ? reuse_spl_n : alg.n_iters
-
-    # Init samples
-    samples = Vector{Sample}(undef, n)
-    weight = 1 / n
-    for i = 1:n
-        samples[i] = Sample(weight, Dict{Symbol, Any}())
-    end
-
-    # Create VarInfo
-    vi = if resume_from == nothing
-        VarInfo(model)
-    else
-        deepcopy(resume_from.info[:vi])
-    end
-
-    # Get `init_theta`
-    if init_theta != nothing
-        verbose && @info "Using passed-in initial variable values" init_theta
-        # Convert individual numbers to length 1 vector; `ismissing(v)` is needed as `size(missing)` is undefined`
-        init_theta = [ismissing(v) || size(v) == () ? [v] : v for v in init_theta]
-        # Flatten `init_theta`
-        init_theta_flat = foldl(vcat, map(vec, init_theta))
-        # Create a mask to inidicate which values are not missing
-        theta_mask = map(x -> !ismissing(x), init_theta_flat)
-        # Get all values
-        theta = vi[spl]
-        @assert length(theta) == length(init_theta_flat) "Provided initial value doesn't match the dimension of the model"
-        # Update those which are provided (i.e. not missing)
-        theta[theta_mask] .= init_theta_flat[theta_mask]
-        # Update in `vi`
-        vi[spl] = theta
-    end
-
-    # Convert to transformed space
-    if spl.selector.tag == :default
-        link!(vi, spl)
-        runmodel!(model, vi, spl)
-    end
-
-    # Init h, prop and adaptor
-    step(model, spl, vi, Val(true); rng=rng, adaptor=adaptor)
-
-    # Sampling using AHMC and store samples in `samples`
-    steps!(model, spl, vi, samples; rng=rng, verbose=verbose, progress=progress)
+    s::Selector=Selector()
+) where {SamplerType<:Union{StaticHamiltonian, AdaptiveHamiltonian}}
+    info = Dict{Symbol, Any}()
+    state_bad = SamplerState(VarInfo(model))
```

Comment:
What does thie variable name mean?

Reply:
It's basically a state that is not intended to make it to the fully typed sampler. I do this in two stages to make sure all the AHMC stuff is type stable before it makes it to the full set of interface calls, because AHMC needs to have a `Sampler` at some point to decide what the initial epsilon should look like, among other things.

Once the entire AHMC sampler is fully typed, I make a new sampler and give it the correct state variable.

Reply:
Maybe change the name to something more informative then?

Reply:
Any plan to rename them?